### PR TITLE
JAMES-4086 Mailet for folding long lines

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: docs/modules/servers/partials/FoldLongLines.adoc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+=== FoldLongLines
+
+This mailet fold (wrap) any header lines of the mail that exceed the maximum number of characters.
+
+It takes only one parameter:
+
+* maxCharacters: maximum number of characters. Default to 998.
```
Comment (quantranhong1999) [inline docs/modules/servers/partials/FoldLongLines.adoc:7]: `maximum number of characters per line`? Can we note why it is 998 BTW?)

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/MimeUtil.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+public class MimeUtil {
+    /**
+     * Splits the specified string into a multiple-line representation with
+     * lines no longer than the maximum number of characters (because the line might contain
+     * encoded words; see <a href='http://www.faqs.org/rfcs/rfc2047.html'>RFC
+     * 2047</a> section 2). If the string contains non-whitespace sequences
+     * longer than the maximum number of characters a line break is inserted at the whitespace
+     * character following the sequence resulting in a line longer than 76
+     * characters.
```
Comment (quantranhong1999) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/MimeUtil.java:30]: not 76 chars anymore?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import jakarta.mail.Header;
+import jakarta.mail.MessagingException;
+
+import org.apache.commons.lang3.stream.Streams;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMailet;
+
+import com.github.fge.lambdas.Throwing;
+
+public class FoldLongLines extends GenericMailet {
```
Comment (quantranhong1999) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:33]: Can we have a little JavaDoc too?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            .name("mail").mimeMessage(MimeMessageBuilder.mimeMessageBuilder().addHeader(HEADER_NAME, HEADER_VALUE).build())
+            .build();
+        foldMailet.service(mail);
+
+        String actual = Streams.of(mail.getMessage().getAllHeaders()).filter(header -> header.getName().equals(HEADER_NAME)).findFirst().get().getValue();
+        assertThat(actual).isEqualTo(FOLDED_LINE);
+    }
+
+    @Test
+    void serviceShouldNotFoldLinesWhenTheyExceedMaxCharacters() throws MessagingException {
```
Comment (quantranhong1999) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java:82]:
```suggestion
    void serviceShouldNotFoldLinesWhenTheyNotExceedMaxCharacters() throws MessagingException {
```

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            .setProperty(MAX_CHARACTERS_PARAMETER_NAME, "30")
+            .build();
+        foldMailet.init(mailetConfig);
+
+        Mail mail = FakeMail.builder()
+            .name("mail").mimeMessage(MimeMessageBuilder.mimeMessageBuilder().addHeader(HEADER_NAME, HEADER_VALUE).build())
+            .build();
+        foldMailet.service(mail);
+
+        String actual = Streams.of(mail.getMessage().getAllHeaders()).filter(header -> header.getName().equals(HEADER_NAME)).findFirst().get().getValue();
```
Comment (vttranlina) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java:77]: verify 
`mail.getMessage().getAllHeaders()).filter(header -> header.getName().equals(HEADER_NAME)` has size = `1` 
it is more strong, 
This strengthens the validation and ensures we replace the header value instead of appending a new one.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        List<Header> foldedHeaders = Streams.of(mail.getMessage().getAllHeaders().asIterator()).map(this::fold).toList();
+        foldedHeaders.forEach(Throwing.consumer(header -> mail.getMessage().setHeader(header.getName(), header.getValue())));
+    }
+
+    private Header fold(Header header) {
+        // TODO After new release of mime4j, update to use MimeUtil of mime4j and remove MimeUtil class file
+        String folded = MimeUtil.fold(header.getValue(), header.getName().length() + 2, maxCharacters);
```
Comment (vttranlina) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:61]: Why `2` ? 
this is for ": " ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    private Integer maxCharacters;
+
+    @Override
+    public void init() throws MessagingException {
+        maxCharacters = getInitParameterAsOptional(MAX_CHARACTERS_PARAMETER_NAME).map(Integer::parseInt).orElse(MAX_CHARACTERS);
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        List<Header> foldedHeaders = Streams.of(mail.getMessage().getAllHeaders().asIterator()).map(this::fold).toList();
```
Comment (vttranlina) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:55]: IMO, filter `maxCharacters` here is well
Then we don't need care the logic of `MimeUtil`

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Override
+    public void init() throws MessagingException {
+        maxCharacters = getInitParameterAsOptional(MAX_CHARACTERS_PARAMETER_NAME).map(Integer::parseInt).orElse(DEFAULT_MAX_CHARACTERS);
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        List<Header> foldedHeaders = Streams.of(mail.getMessage().getAllHeaders().asIterator()).map(this::fold).toList();
+        foldedHeaders.forEach(Throwing.consumer(header -> mail.getMessage().setHeader(header.getName(), header.getValue())));
```
[the remark this answers is not in the document]
Comment (hungphan227) [reply mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:56]: I believe so

[the remark this answers is not in the document]
Comment (hungphan227) [reply mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:56]: doc of setHeader:
```

     * Change the first header line that matches name
     * to have value, adding a new header if no existing header
     * matches. Remove all matching headers but the first. <p>
```

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java (hunk 7 above)
Comment (hungphan227) [reply mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:55]: IMO it does not really matter

File: docs/modules/servers/partials/FoldLongLines.adoc [hunk 9]
```diff
@@ -0,0 +1,7 @@
+=== FoldLongLines
+
+This mailet fold (wrap) any header lines of the mail that exceed the maximum number of characters.
+
+It takes only one parameter:
+
+* maxCharacters: maximum number of characters. Default to 998 (This line limit is defined in RFC5322).
```
Comment (chibenwa) [inline docs/modules/servers/partials/FoldLongLines.adoc:7]: Please explain more:

```
This sanitizing of EML message will prevent downstream server from applying it, and prevent them to break the DKIM signature positionned by James.
```

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+public class FoldLongLines extends GenericMailet {
+    public static final String MAX_CHARACTERS_PARAMETER_NAME = "maxCharacters";
+
+    private static final int DEFAULT_MAX_CHARACTERS = 998;
+
+    private Integer maxCharacters;
+
+    @Override
+    public void init() throws MessagingException {
+        maxCharacters = getInitParameterAsOptional(MAX_CHARACTERS_PARAMETER_NAME).map(Integer::parseInt).orElse(DEFAULT_MAX_CHARACTERS);
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:50]: Negative value is ok? 0 is ok?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.mailet.Mailet;
+import org.apache.mailet.MailetContext;
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMailContext;
+import org.apache.mailet.base.test.FakeMailetConfig;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+
+public class FoldLongLinesTest {
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java:43]: I am missing edge case tests eg when I have twic the same header name:

```
Received: abc
Received: def
```

=> Resulting solution shall ensure the two headers are still present, I'm not sure it's the case today

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java (hunk 8 above)
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:56]: I would really love that we do mime modification only if we need to.

So if the mail contains no long headers, ie FoldLongLines noops, then we do not attempt any mime modifications.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        List<Header> foldedHeaders = Streams.of(mail.getMessage().getAllHeaders().asIterator()).map(this::fold).toList();
+        foldedHeaders.forEach(Throwing.consumer(header -> mail.getMessage().setHeader(header.getName(), header.getValue())));
+    }
+
+    private Header fold(Header header) {
+        // TODO After new release of mime4j, update to use MimeUtil of mime4j and remove MimeUtil class file
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:60]: Link a PR, a commit on master, a ticket number, anything...

**Review by chibenwa — CHANGES_REQUESTED** [review]
Unsufficiently tested changeset that is actually dangerous as it is.

**Comment by chibenwa** [comment]
Let's add an integretion test

```
GIVEN bob sends a mail to ALICE with long headers
WHEN alice checks her mails in her INBOX
THEN the headers are folded
AND the DKIM signature is correct
```

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 13]
```diff
@@ -40,25 +44,45 @@
  */
 public class FoldLongLines extends GenericMailet {
     public static final String MAX_CHARACTERS_PARAMETER_NAME = "maxCharacters";
+    public static final String HEADER_SEPARATOR = ": ";
 
     private static final int DEFAULT_MAX_CHARACTERS = 998;
+    private static final String SEPARATE_LINE_REGEX = "\r\n|\n";
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:50]: Please let's not use REGEX

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 14]
```diff
@@ -40,25 +44,45 @@
...
+        Preconditions.checkArgument(maxCharacters > 0, "maxCharacters must be positive");
+        this.maxCharacters = maxCharacters;
     }
 
     @Override
     public void service(Mail mail) throws MessagingException {
-        List<Header> foldedHeaders = Streams.of(mail.getMessage().getAllHeaders().asIterator()).map(this::fold).toList();
-        foldedHeaders.forEach(Throwing.consumer(header -> mail.getMessage().setHeader(header.getName(), header.getValue())));
+        // get all header names of headers that exceed line limit
+        Set<String> longHeaders = Streams.of(mail.getMessage().getAllHeaders().asIterator()).filter(this::exceedLineLimit).map(Header::getName).collect(ImmutableSet.toImmutableSet());
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:64]: Extract to a submethod and indent

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 15]
```diff
@@ -40,25 +44,45 @@
...
+        // get all headers that have the same name as above headers
+        List<Header> headers = Streams.of(mail.getMessage().getAllHeaders().asIterator()).filter(header -> longHeaders.contains(header.getName())).toList();
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:66]: Extract to a sub method and indent

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 16]
```diff
@@ -40,25 +44,45 @@
...
+        mail.getMessage().saveChanges();
     }
 
-    private Header fold(Header header) {
-        // TODO After new release of mime4j, update to use MimeUtil of mime4j and remove MimeUtil class file
-        String folded = MimeUtil.fold(header.getValue(), header.getName().length() + 2, maxCharacters);
-        return new Header(header.getName(), folded);
+    private String fold(Header header) {
+        // TODO After new release of mime4j with commit https://github.com/apache/james-mime4j/commit/66a09219457854c7a26e5b7c0e4c9dd59b4b0c32, update to use MimeUtil of mime4j and remove MimeUtil class file
+        return MimeUtil.fold(header.getValue(), header.getName().length() + HEADER_SEPARATOR.length(), maxCharacters);
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:81]:
```suggestion
        int headerNameLength = header.getName().length() + HEADER_SEPARATOR.length();
        return MimeUtil.fold(header.getValue(), header.getName().length() + HEADER_SEPARATOR.length(), maxCharacters);
```

Not easy to guess!

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 17]
```diff
@@ -40,25 +44,45 @@
...
+    }
+
+    private boolean exceedLineLimit(Header header) {
+        String fullHeader = header.getName() + HEADER_SEPARATOR + header.getValue();
+        return Arrays.stream(fullHeader.split(SEPARATE_LINE_REGEX)).anyMatch(line -> line.length() > maxCharacters);
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:86]: Use Splitter.on(...) instead.

\n is enough

**Review by Arsnael — COMMENTED** [review]
Read it nothing to add

**Comment by hungphan227** [comment]
> Let's add an integretion test
> 
> ```
> GIVEN bob sends a mail to ALICE with long headers
> WHEN alice checks her mails in her INBOX
> THEN the headers are folded
> AND the DKIM signature is correct
> ```

Added

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 18]
```diff
@@ -73,16 +72,33 @@ public void service(Mail mail) throws MessagingException {
...
+            .filter(header -> longHeaders.contains(header.getName()))
+            .toList();
     }
 
     private String fold(Header header) {
+        int headerNameLength = header.getName().length() + HEADER_SEPARATOR.length();
         // TODO After new release of mime4j with commit https://github.com/apache/james-mime4j/commit/66a09219457854c7a26e5b7c0e4c9dd59b4b0c32, update to use MimeUtil of mime4j and remove MimeUtil class file
-        return MimeUtil.fold(header.getValue(), header.getName().length() + HEADER_SEPARATOR.length(), maxCharacters);
+        String s = MimeUtil.fold(header.getValue(), headerNameLength, maxCharacters);
+        return s;
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:97]:
```suggestion
        return MimeUtil.fold(header.getValue(), headerNameLength, maxCharacters);
```

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 19]
```diff
@@ -73,16 +72,33 @@ public void service(Mail mail) throws MessagingException {
...
     }
 
     private boolean exceedLineLimit(Header header) {
         String fullHeader = header.getName() + HEADER_SEPARATOR + header.getValue();
-        return Arrays.stream(fullHeader.split(SEPARATE_LINE_REGEX)).anyMatch(line -> line.length() > maxCharacters);
+        return Splitter.on(SEPARATOR).splitToStream(fullHeader).anyMatch(line -> line.length() > maxCharacters);
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:102]: Ident?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/MimeUtil.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets;
+
+public class MimeUtil {
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/MimeUtil.java:22]: We can put a todo comment in this class too

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        headers.forEach(Throwing.consumer(header -> {
+            if (exceedLineLimit(header)) {
+                mail.getMessage().addHeader(header.getName(), fold(header));
+            } else {
+                mail.getMessage().addHeader(header.getName(), header.getValue());
+            }
+        }));
+        if (!longHeaders.isEmpty()) {
+            mail.getMessage().saveChanges();
+        }
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:77]:
```
        Set<String> longHeaders = getHeadersExceedingMaxCharacters(mail);
        if (1longHeaders.isEmpty()) {
             // all other stuff
         }
```

I insist!

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+    public static final String HEADER_SEPARATOR = ": ";
+
+    private static final int DEFAULT_MAX_CHARACTERS = 998;
+    private static final String SEPARATOR = "\n";
+
+    private Integer maxCharacters;
+
+    @Override
+    public void init() throws MessagingException {
+        Integer maxCharacters = getInitParameterAsOptional(MAX_CHARACTERS_PARAMETER_NAME).map(Integer::parseInt).orElse(DEFAULT_MAX_CHARACTERS);
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:56]:
```suggestion
        int maxCharacters = getInitParameterAsOptional(MAX_CHARACTERS_PARAMETER_NAME).map(Integer::parseInt).orElse(DEFAULT_MAX_CHARACTERS);
```

It is not allowed to be null, right?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+ * </ul>
+ */
+public class FoldLongLines extends GenericMailet {
+    public static final String MAX_CHARACTERS_PARAMETER_NAME = "maxCharacters";
+    public static final String HEADER_SEPARATOR = ": ";
+
+    private static final int DEFAULT_MAX_CHARACTERS = 998;
+    private static final String SEPARATOR = "\n";
+
+    private Integer maxCharacters;
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/FoldLongLines.java:52]: Use `int`: it is not allowed to be null

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+
+    private Mailet foldMailet;
+    private MailetContext mailetContext;
+
+    @BeforeEach
+    void beforeEach() {
+        foldMailet = new FoldLongLines();
+        mailetContext = FakeMailContext.builder()
+            .logger(mock(Logger.class))
+            .build();
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java:58]:
```suggestion
        mailetContext = FakeMailContext.defaultContext();
```

?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+                .addHeader(HEADER_NAME, HEADER_VALUE)
+                .build()
+            )
+            .build();
+        foldMailet.service(mail);
+
+        List<Header> headers = Streams.of(mail.getMessage().getAllHeaders()).filter(header -> header.getName().equals(HEADER_NAME)).toList();
+        assertThat(headers).hasSize(2);
+        assertThat(headers.getFirst().getValue()).isEqualTo("<b1@gmailcom>");
+        assertThat(headers.getLast().getValue()).isEqualTo(FOLDED_LINE);
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/FoldLongLinesTest.java:160]: Wrap this in soft assertions?

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]
